### PR TITLE
P2: LILA-5460: Add OffsetHelper to SDK

### PR DESCRIPTION
**Pull request by GigaHierz**
- [ ] add OffsetHelper, 
- [ ] restrict ETH functions for Celo

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+
+fetchProjectById();
+
+// Fetch all aggregations (including for example, tco2TotalRetired or totalCarbonBridged).
+fetchAggregations();
+```
+
+## Custom queries
+
+There's a lot more other pre-built subgraph queries that I could show you, but what I really want to show you is the `fetchCustomQuery` method.
```
Comment (pheuberger) [inline docs/README.md:176]: "There are a lot more pre-built..."

File: docs/README.md [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+This allows you to fetch with your own queries and can be very powerful if you know graphQL. You can also check out a lot of example queries in our subgraph [playgrounds](https://thegraph.com/hosted-service/subgraph/toucanprotocol/matic).
+
+- Getting all infos on a project of a Carbon Credit
```
Comment (pheuberger) [inline docs/README.md:180]: "[...] on a carbon project"

Comment (pheuberger) [reply docs/README.md:180]: Might also make sense to tell people that `region` is not actually the region but the country :smiling_face_with_tear:

File: docs/README.md [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+        id
+      }
+    }
+  }
+`;
+
+const result = await toucan.fetchCustomQuery(query, { id: "1" });
+```
+
+# AutoOffset using the OffsetHelper
```
Comment (pheuberger) [inline docs/README.md:202]: Is `AutoOffset` a symbol name or why is it PascalCase'd like that?

File: docs/README.md [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+Using the OffsetHelper methods is the easiest way to offset CO2 right now as it handles the whole process for you.
+
+_(Uses Ubeswap/SushiSwap to acquire BCT/NCT, redeems it for TCO2s and retires them)_
+
+This is how you do that when you want to use ETH (or the network's native currency) in your balance.
+
+In this example, you'd be offsetting 1 TON of CO2 with carbon projects from the NCT pool.
+
+```typescript
+const offsetReceipt = await toucan.autoOffsetUsingETH("NCT", parseEther("1.0"));
```
Comment (pheuberger) [inline docs/README.md:217]: Just want to confirm if I understand the interface correctly. You're specifying how many tons of NCT you want and it will calculate the required ETH price for you?

File: docs/README.md [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+Of course, you could have avoided using the `getScoredTCO2s` method if you already had the address of the TCO2s you wanted to offset.
+
+### But I want the Retirement NFT
+
+If you do, that's also quite easy to pull off. You just replace the usage of the `retire()` method in the example above with this:
+
+```typescript
+const retirementReceipt = await toucan.retireAndMintCertificate(
+  "Alex",
```
Comment (pheuberger) [inline docs/README.md:277]: Let's use Alice here and below too. It's the standard way of naming stuff like this :)

File: docs/README.md [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+  signer.address,
+  "Alex",
+  "Just helping the planet",
+  parseEther("3.0"),
+  tco2s[len - 1].address
+);
+```
+
+Why do you see my name twice you ask? 🤔
```
Comment (pheuberger) [inline docs/README.md:286]: Use Alice here as well

File: docs/README.md [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+
+Well, the first "Alex" represents the entity that is doing the retirement/offset. The second one represents the party that 'benefits' from it. The address also represents the beneficiary.
```
Comment (pheuberger) [inline docs/README.md:288]: Again, Alice

Comment (pheuberger) [reply docs/README.md:288]: I think we can assume people understand the term beneficiary, so I'd reword the part about the party benefitting from it. Also, which address represents the beneficiary? I can only see the signer address.

File: docs/classes/default.md [hunk 8]
```diff
@@ -0,0 +1,1281 @@
+[toucan-sdk](../README.md) / [Exports](../modules.md) / default
```
Comment (pheuberger) [inline docs/classes/default.md:1]: Should I review this entire file as well?

File: docs/modules.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
+[toucan-sdk](README.md) / Exports
```
Comment (pheuberger) [inline docs/modules.md:1]: Should I review this entire file as well?

File: package.json [hunk 10]
```diff
@@ -37,7 +39,6 @@
     "@typechain/hardhat": "^6.0.0",
     "@types/chai": "^4.2.21",
     "@types/mocha": "^9.0.0",
-    "@types/node": "^17.0.34",
```
Comment (pheuberger) [inline package.json:40]: node types are not needed anymore?

File: src/.DS_Store
Comment (pheuberger) [inline src/.DS_Store:1]: This file shouldn't be checked in. Please remove and add to `.gitignore`

File: src/index.ts
Comment (pheuberger) [inline src/index.ts:1]: Oof, that's a big one :see_no_evil: 

I'll skip for now and attempt in a second sitting.

File: src/subclasses/ContractInteractions.ts [hunk 11]
```diff
@@ -33,19 +33,20 @@ import {
 import { Network, PoolSymbol } from "../types";
 import { GAS_LIMIT } from "../utils";
 import {
+  offsetHelperABI,
   poolTokenABI,
   tco2ABI,
   toucanContractRegistryABI,
 } from "../utils/ABIs";
-import addresses, { IfcOneNetworksAddresses } from "../utils/addresses";
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:40]: What does "Ifc" stand for? Interface?

File: src/subclasses/ContractInteractions.ts [hunk 12]
```diff
@@ -64,7 +65,9 @@ class ContractInteractions {
 
   /**
    *
-   * @description retires/burns an amount of TCO2s (each represents 1 ton of CO2) to achieve offset
+   * @description retires/burns an amount of TCO2s (each represents 1 ton of CO2)
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:68]: "to achieve offset" is a strange wording. Can you reword?

File: src/subclasses/ContractInteractions.ts [hunk 13]
```diff
@@ -417,14 +412,16 @@ class ContractInteractions {
 
   /**
    *
-   * @description allows user to retire carbon using carbon pool tokens from his wallet
-   * @notice this method may take up to even 1 minute to give a result
-   * @param pool symbol of the pool (token) to use
-   * @param amount amount of CO2 tons to offset
+   * @description Retire carbon credits using the lowest quality (oldest) TCO2
+   * tokens available by sending Toucan pool tokens,  e.g., NCT.
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:416]: Double space in between comma and "e.g."

File: src/subclasses/ContractInteractions.ts [hunk 14]
```diff
@@ -417,14 +412,16 @@ class ContractInteractions {
...
+   * @notice This method may take up to even 1 minute to give a result
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:417]: scratch even and replace "give" with "return"

File: src/subclasses/ContractInteractions.ts [hunk 15]
```diff
@@ -417,14 +412,16 @@ class ContractInteractions {
...
+   * @param pool symbol of the pool token to offset,
+   *  e.g., NCT
+   * @param amount amount of of TCO2 to offset.
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:420]: "of" twice here. Would also add "The" so amount isn't directly repeated

File: src/subclasses/ContractInteractions.ts [hunk 16]
```diff
@@ -440,26 +437,27 @@ class ContractInteractions {
...
         gasLimit: GAS_LIMIT,
       });
     return await offsetTxn.wait();
   };
 
   /**
    *
-   * @description swaps given token for carbon pool tokens and uses them to retire carbon
-   * @notice this method may take up to even 1 minute to give a result
+   * @description Retire a specified amount of  carbon credits using the lowest quality (oldest) TCO2 tokens available from the specified Toucan token pool by sending ERC20
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:448]: Double space between "of" and "carbon"

Comment (pheuberger) [reply src/subclasses/ContractInteractions.ts:448]: "Toucan token" is probably not a good wording as that might be the name of our protocol token.

File: src/subclasses/ContractInteractions.ts [hunk 17]
```diff
@@ -440,26 +437,27 @@ class ContractInteractions {
...
+   * tokens (cUSD, USDC, WETH, WMATIC).
+   * @notice This method needs two different actions signed and may take up to even 1 minute to give a result
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:450]: Same as above, even and give

File: src/subclasses/ContractInteractions.ts [hunk 18]
```diff
@@ -440,26 +437,27 @@ class ContractInteractions {
...
+   * @param swapToken portal for the token to swap into pool tokens (only accepts WETH, WMATIC and USDC)
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:451]: Is `swapToken` a good name for this?

File: src/subclasses/ContractInteractions.ts [hunk 19]
```diff
@@ -440,26 +437,27 @@ class ContractInteractions {
...
    * @param pool symbol of the pool (token) to use
    * @param amount amount of CO2 tons to offset
-   * @param swapToken portal for the token to swap into pool tokens (only accepts WETH, WMATIC and USDC)
    * @param signer this being a write transaction, we need a signer
    * @returns offset transaction
    */
-  autoOffsetUsingSwapToken = async (
+  autoOffsetExactOutToken = async (
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:457]: I don't understand the function name. Can you explain?

File: src/subclasses/ContractInteractions.ts [hunk 20]
```diff
@@ -487,14 +486,55 @@ class ContractInteractions {
 
   /**
    *
-   * @description swaps ETH for carbon pool tokens and uses them to retire carbon
-   * @notice this method may take up to even 1 minute to give a result
+   * @description retire  carbon credits using the lowest quality (oldest) TCO2 tokens available from the specified Toucan token pool by sending ERC20
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:489]: Double space in between retire and carbon.

File: src/subclasses/ContractInteractions.ts [hunk 21]
```diff
@@ -487,14 +486,55 @@ class ContractInteractions {
...
+   * tokens (cUSD, USDC, WETH, WMATIC). All provided token is consumed for
+   * offsetting.
+   * @notice This method needs two different actions signed and may take up to even 1 minute to give a result
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:492]: Same as above, even and give

File: src/subclasses/ContractInteractions.ts [hunk 22]
```diff
@@ -487,14 +486,55 @@ class ContractInteractions {
 
   /**
    *
-   * @description swaps ETH for carbon pool tokens and uses them to retire carbon
-   * @notice this method may take up to even 1 minute to give a result
+   * @description retire  carbon credits using the lowest quality (oldest) TCO2 tokens available from the specified Toucan token pool by sending ERC20
+   * tokens (cUSD, USDC, WETH, WMATIC). All provided token is consumed for
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:490]: [...] "tokens are" [...]

File: src/subclasses/ContractInteractions.ts [hunk 23]
```diff
@@ -487,14 +486,55 @@ class ContractInteractions {
...
+   * offsetting.
+   * @notice This method needs two different actions signed and may take up to even 1 minute to give a result
+   *  @dev When automatically redeeming pool tokens for the lowest quality
+   * TCO2s there are no fees and you receive exactly 1 TCO2 token for 1 pool
+   * token.
+   * @param swapToken portal for the token to swap into pool tokens (only accepts WETH, WMATIC and USDC)
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:496]: Again, `swapToken` variable name is confusing to me

File: src/subclasses/ContractInteractions.ts [hunk 24]
```diff
@@ -487,14 +486,55 @@ class ContractInteractions {
...
+   * @param pool symbol of the pool (token) to use
+   * @param amount the amount of ERC20 token to swap into Toucan pool token. Full amount will be used for offsetting.
+   * @param signer this being a write transaction, we need a signer
+   * @returns offset transaction
+   */
+  autoOffsetExactInToken = async (
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:502]: Same question about function name.

File: src/subclasses/ContractInteractions.ts [hunk 25]
```diff
@@ -487,14 +486,55 @@ class ContractInteractions {
...
+        poolAddress,
+        amount,
+        { gasLimit: GAS_LIMIT }
+      );
+    return await offsetTxn.wait();
+  };
+
+  /**
+   *
+   * @description Retire a specified amount of  carbon credits using the lowest quality (oldest) TCO2 tokens available from the specified Toucan token pool by sending a native token e.g. MATIC.
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:529]: Double space in between "of" and "carbon" and "Toucan token" again

File: src/subclasses/ContractInteractions.ts [hunk 26]
```diff
@@ -487,14 +486,55 @@ class ContractInteractions {
...
+   * @dev Use `calculateNeededETHAmount()` first in order to find out how much  of the native token e.g. MATIC is required to retire the specified quantity of TCO2. If the user sends much native token e.g. MATIC, the leftover amount will be sent back to the user.
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:530]: Double space in between "much" and "of"

File: src/subclasses/ContractInteractions.ts [hunk 27]
```diff
@@ -512,17 +552,46 @@ class ContractInteractions {
 
   /**
    *
-   * @description calculates the needed amount of tokens to send to offset
-   * @param pool symbol of the pool (token) to use
-   * @param amount amount of CO2 tons to calculate for
-   * @param swapToken contract of the token to use in swap
+   * @description swaps ETH for carbon pool tokens and uses them to retire carbon
+   * @notice This method may take up to even 1 minute to give a result
```
Comment (pheuberger) [inline src/subclasses/ContractInteractions.ts:556]: Even and give

**Review by pheuberger — DISMISSED** [review]
There's a lot of duplicate things I mentioned. I didn't mention all of them. Just do a project wide search and change all of these :) I didn't review everything yet. Just sending this so you have something to go off of

File: README.md [hunk 28]
```diff
@@ -222,7 +222,7 @@ You can use different tokens to do it. Currently it works with USDC, WMATIC and
 ```typescript
 const weth = new ethers.Contract(wethAddress, wethAbi, signer);
 
-const offsetReceipt = await toucan.autoOffsetUsingSwapToken(
+const offsetReceipt = await toucan.autoOffsetUsingToken(
```
[the remark this answers is not in the document]
Comment (GigaHierz) [reply README.md:225]: the OffsetHelper has already been updated but the SDK was still based on the old versions. I changed the naming so it would fit the naming of the offsetHelper functions. But this is actually wrong update of the function name. Correct would be `autoOffsetExactInToken` . they had to be changed because before the function didn't differentiate between `autoOffsetExactInToken` and `autoOffsetExactOutToken`

File: docs/README.md [hunk 29]
```diff
@@ -0,0 +1,290 @@
...
+    }
+  }
+}
+```
+
+Now you have quite some info on the project, including its address.
+
+## All queries:
+
+Toucan SDK offers a lot of pre-defined queries. Try them out!
```
[the remark this answers is not in the document]
Comment (GigaHierz) [reply docs/README.md:138]: I am not sure what you mean by incomplete query syntax. in the custom query example I am using the full syntax. The the functions here are the SDK functions. So, there is not really a query syntax.

I've added skip and first where it makes sense. 
TBH, the only thing that was updated for the subgraph queries, was adding Celo and Alfajores. But already happened in the last commit.

File: docs/README.md (hunk 4 above)
Comment (GigaHierz) [reply docs/README.md:217]: Yes, and thank you for flaggin this as the function name was also updated here to `autoOffsetExactOutETH`. You would use `autoOffsetExactOutETH` if you would want to offset e.g., 5 USDC

File: docs/classes/default.md (hunk 8 above)
Comment (GigaHierz) [reply docs/classes/default.md:1]: nope

File: docs/modules.md (hunk 9 above)
Comment (GigaHierz) [reply docs/modules.md:1]: nope

File: package.json (hunk 10 above)
Comment (GigaHierz) [reply package.json:40]: it was giving dependency errors, so I removed it. And that didn't change anything for any operations e.g., tests etc., so apparently it's not needed.

File: src/subclasses/ContractInteractions.ts (hunk 11 above)
Comment (GigaHierz) [reply src/subclasses/ContractInteractions.ts:40]: yes.

File: src/subclasses/ContractInteractions.ts (hunk 19 above)
Comment (GigaHierz) [reply src/subclasses/ContractInteractions.ts:457]: Retiring carbon tokens requires multiple steps and interactions with Toucan Protocol's main contracts:

- Obtain a Toucan pool token such as BCT or NCT (by performing a token swap).
- Redeem the pool token for a TCO2 token.
- Retire the TCO2 token.

These steps are combined in each of the following "auto offset" methods implemented in OffsetHelper to allow a retirement within one transaction:

`autoOffsetPoolToken()` if the user already owns a Toucan pool token such as BCT or NCT,
`autoOffsetExactOutETH()` if the user would like to perform a retirement using MATIC, specifying the exact amount of TCO2s to retire,
`autoOffsetExactInETH()` if the user would like to perform a retirement using MATIC, swapping all sent MATIC into TCO2s,
`autoOffsetExactOutToken()` if the user would like to perform a retirement using an ERC20 token (USDC, WETH or WMATIC), specifying the exact amount of TCO2s to retire,
`autoOffsetExactInToken()` if the user would like to perform a retirement using an ERC20 token (USDC, WETH or WMATIC), specifying the exact amount of token to swap into TCO2s.

File: docs/README.md [hunk 30]
```diff
@@ -0,0 +1,290 @@
+toucan-sdk / [Exports](modules.md)
+
+# Toucan SDK
+
+[![NPM Package](https://img.shields.io/npm/v/toucan-sdk.svg)](https://www.npmjs.com/package/toucan-sdk)
+
+A Typescript SDK for [Toucan Protocol](https://toucan.earth/). Wraps around [Ethers.js](https://docs.ethers.io/v5/). Works in the web browser and Node.js.
```
[the remark this answers is not in the document]
Comment (GigaHierz) [reply docs/README.md:9]: These docs also need to be added to the .gitignore file as they automatically created. Just leaving them in right now till all conversations are resolved as I've updated these in the main README.md file.

**Review by Niwilai — COMMENTED** [review]
I need to submit current questions and continue with the review

File: src/subclasses/ContractInteractions.ts [hunk 31]
```diff
@@ -470,13 +467,14 @@ class ContractInteractions {
       await offsetHelper.calculateNeededTokenAmount(
         swapToken.address,
         poolAddress,
-        amount
+        amount,
+        { gasLimit: GAS_LIMIT }
```
[the remark this answers is not in the document]
Comment (GigaHierz) [reply src/subclasses/ContractInteractions.ts]: I think I've added it when I was debugging and didn't remove it

File: src/subclasses/ContractInteractions.ts (hunk 19 above)
Comment (GigaHierz) [reply src/subclasses/ContractInteractions.ts:451]: @Niwilai the function names relate to swap functions that have the same naming. In both cases you need to swap. 
The difference is that for `OutETH` means you are specifying the exact amount of TCO2s to retire. for this function you will have to call `calculateNeededETHAmount` to know how many tokens you need to send. 

In means that you define the amount of tokens you want to swap for pool tokens. Here the amount you want to pay is more important than the amount of TCO2 you offset. useful for a subscription where you would want to retire 5$ per month. here you can use the `calculateExpectedPoolTokenForETH` to calculate beforehand how many TCO2 s you will get back/ you can retire.

File: src/subclasses/ContractInteractions.ts (hunk 23 above)
Comment (GigaHierz) [reply src/subclasses/ContractInteractions.ts:487]: it's being used here as that is the token that will be swapped for pool tokens

File: docs/README.md [hunk 32]
```diff
@@ -0,0 +1,299 @@
...
+```
+
+# Quickstart
+
+Instantiate the ToucanClient and set a `signer` & `provider` to interact with our infrastructure.
+
+```typescript
+import { ToucanClient } from "toucan-sdk";
+
+const toucan = new ToucanClient("alfajores", provider, signer);
```
Comment (Niwilai) [inline docs/README.md:34]: I think it would be great to link how to instantiate a provider / signer for Ethers or if using something like Wagmi, to import the signer/provider

File: src/subclasses/ContractInteractions.ts [hunk 33]
```diff
@@ -476,25 +468,63 @@ class ContractInteractions {
     await approveTxn.wait();
 
     const offsetTxn: ContractTransaction =
-      await offsetHelper.autoOffsetUsingToken(
+      await offsetHelper.autoOffsetExactOutToken(
         swapToken.address,
         poolAddress,
-        amount,
-        { gasLimit: GAS_LIMIT }
```
Comment (Niwilai) [inline src/subclasses/ContractInteractions.ts:483]: You didn't remove them consistently across all methods, however on a second thought, I think there might be benefit to keep it so that we don't run into reversed txs. So sorry for withdrawing my first review point 🙏

File: src/subclasses/ContractInteractions.ts [hunk 34]
```diff
@@ -636,7 +746,13 @@ class ContractInteractions {
   public getOffsetHelperContract = (
     signerOrProvider: ethers.Signer | ethers.providers.Provider
   ): OffsetHelper => {
-    throw new Error("OffsetHelper is not supported yet");
+    // throw new Error("OffsetHelper is not supported yet");
```
Comment (Niwilai) [inline src/subclasses/ContractInteractions.ts:749]: Any reason to leave this in :) ?

**Review by Niwilai — COMMENTED** [review]
Looks good, but very hard to review due to a big diff. Some comments:

File: docs/README.md (hunk 32 above)
Comment (GigaHierz) [reply docs/README.md:34]: added to the main readme

File: test/index.test.ts [hunk 35]
```diff
@@ -146,46 +144,95 @@ describe("Testing Toucan-SDK contract interactions", function () {
...
+        formatEther(state[0].wethBalance.sub(ONE_ETHER)),
+        `Expect addr1's WETH balance to be less by 1.0`
       ).to.be.equal(formatEther(state[1].wethBalance));
     });
 
-    it("Should retire 1 TCO2 using ETH deposit", async function () {
+    it("Should retire 1 TCO2 using ETH (native token) #autoOffsetExactInETH)", async function () {
+      const pool = await toucan.getPoolContract("NCT");
+
+      const state: any[] = [];
```
[the remark this answers is not in the document]
Comment (GigaHierz) [reply test/index.test.ts]: I've added some typing but had to keep an any otherwise throws some errors and I haven't been able to figure out how to solve this. pls check. maybe you have an idea. otherwise, can we please get this closed?

**Review by Niwilai — APPROVED** [review]
Approving here to wish you a farewell and don't drag you around. 👋 However, I will do a follow up PR with 1-2 amends that I still found.

**Comment by Niwilai** [comment]
@GigaHierz please rebase 🙏
